src/sentry/integrations/perforce/integration.py

```diff
@@ -302,16 +629,115 @@ def setup(self) -> None:
+                                    "ssl_fingerprint": form_data.get("ssl_fingerprint"),
+                                }
+                            },
+                        )(),
+                        org_integration=type("obj", (object,), {})(),
+                    )
+                    # Test connection by fetching depot list
+                    client.get_depots()
+
+                    pipeline.get_logger().info(
+                        "perforce.setup.connection-verified",
+                        extra={
+                            "p4port": form_data.get("p4port"),
+                            "user": form_data.get("user"),
+                        },
+                    )
+                except ApiUnauthorized as e:
+                    form.add_error(
+                        None,
+                        f"Authentication failed: {e}. Please check your username and password.",
+                    )
+                    return render_to_response(
+                        template="sentry/integrations/perforce-config.html",
+                        context={"form": form},
+                        request=request,
+                    )
+                except ApiError as e:
+                    form.add_error(
+                        None,
+                        f"Failed to connect to Perforce server: {e}. Please verify your server address and SSL fingerprint.",
+                    )
+                    return render_to_response(
+                        template="sentry/integrations/perforce-config.html",
+                        context={"form": form},
+                        request=request,
+                    )
+                except Exception as e:
+                    pipeline.get_logger().error(
+                        "perforce.setup.connection-verification-failed",
+                        extra={
+                            "p4port": form_data.get("p4port"),
+                            "error": str(e),
+                        },
+                        exc_info=True,
+                    )
+                    form.add_error(
+                        None,
+                        f"Unexpected error during connection verification: {e}",
+                    )
+                    return render_to_response(
+                        template="sentry/integrations/perforce-config.html",
+                        context={"form": form},
+                        request=request,
+                    )
+
+                # Bind configuration data to pipeline state
+                pipeline.bind_state("installation_data", form_data)
+                pipeline.bind_state("p4port", form_data.get("p4port"))
+                pipeline.bind_state("name", f"Perforce ({form_data.get('p4port')})")
+                # Include organization_id to create unique external_id per org
+                pipeline.bind_state("organization_id", pipeline.organization.id)
```

Comment:
You can also use `determine_active_organization(request)` when you need to pull the organization in the next step